### PR TITLE
IA-2345: Permissions grouping for IASO admin

### PR DESCRIPTION
**Pull request by hakifran**
Explain what problem this PR is resolving

- Group permissions into modules
- Limit the access of permissions bye modules

Related JIRA tickets : [IA-2345](https://bluesquare.atlassian.net/browse/IA-2345)(Permissions grouping for IASO admin)

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [ ] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [x] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- Added modules and their linked permissions in constants
- Link Account with modules through ArrayField
- Create modules to all existing Accounts
- Add modules JsonField on setup_account
- Edit filter_permissions to take in account modules
- Display on the front end modules(those activated for the account and the other not activated
- Filter on permissions related to current account modules when creating or editing a profile

## How to test
- Go in admin/iaso/account/
- Create an account with some modules
- Create a user with the account
- Check in menu which menus is accessible

## Print screen / video
Screenshot from 2023-10-17 16-17-06
Screenshot from 2023-10-17 16-17-44
f9d98e3
Screenshot from 2023-10-17 16-18-27
)
Screenshot from 2023-10-17 16-19-52

## Notes

Things that the reviewers should know: known bugs that are out of the scope of the PR, other trade-offs that were made.
If the PR depends on a PR in bluesquare-components, or merges into another PR (i.o. main), it should also be mentioned here

[IA-2345]: https://bluesquare.atlassian.net/browse/IA-2345?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by hakifran** [comment]
@cypress

**Comment by hakifran** [comment]
@beygorghor cypress test succeed in local but not on git when launched manually and there is GET /api/entitytypes/?order=name 500 34.038 ms - - on each cypress page test

File: hat/assets/js/apps/Iaso/domains/app/translations/fr.json
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/app/translations/fr.json:1]: Please double check your translations, codename is missing for instance

File: hat/assets/js/apps/Iaso/domains/modules/components/ModulesFilters.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                    type="search"
+                    label={MESSAGES.search}
+                    onEnterPressed={handleSearch}
+                    onErrorChange={setTextSearchError}
+                    blockForbiddenChars
+                />
+            </Grid>
+
+            <Grid
+                container
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/modules/components/ModulesFilters.tsx]: i don't think the container is required here

File: hat/assets/js/apps/Iaso/domains/modules/config.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                Cell: (settings): ReactElement => {
+                    const { account } = settings.row.original;
+                    if (account.length > 0) {
+                        return <CheckedIcon style={{ color: 'green' }} />;
+                    }
+                    return <NotCheckedIcon color="disabled" />;
+                },
+            },
+        ];
+        return columns;
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/modules/config.tsx]: you can directly return the array without using a variable

File: hat/assets/js/apps/Iaso/domains/modules/types/modules.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            // eslint-disable-next-line no-unused-vars
+            data: Array<Module>,
+        ) => Array<any>;
+    };
+
+export type Permission = {
+    id: number;
+    name: string;
+    codename: string;
+};
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/modules/types/modules.ts]: We probably have this one somewhere else

File: hat/assets/js/apps/Iaso/domains/setup/index.tsx [hunk 4]
```diff
@@ -83,6 +85,7 @@ export const SetupAccount: FunctionComponent = () => {
             user_first_name: '',
             user_last_name: '',
             password: '',
+            modules: [],
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/setup/index.tsx:73]: You have to add modules in `useAccountValidation`

File: hat/assets/js/apps/Iaso/domains/setup/index.tsx [hunk 5]
```diff
@@ -102,7 +105,11 @@ export const SetupAccount: FunctionComponent = () => {
     } = formik;
     const onChange = (keyValue, value) => {
         setFieldTouched(keyValue, true);
-        setFieldValue(keyValue, value);
+        if (keyValue === 'modules' && value) {
+            setFieldValue(keyValue, value.split(','));
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/setup/index.tsx]: `commaSeparatedIdsToArray`

File: hat/assets/js/cypress/fixtures/permissions/list.json [hunk 6]
```diff
@@ -57,7 +57,7 @@
         },
         {
             "id": 383,
-            "name": "Unités d'organistations",
```
Comment (beygorghor) [inline hat/assets/js/cypress/fixtures/permissions/list.json:60]: 😇

**Review by beygorghor — CHANGES_REQUESTED** [review]
I have a permission error while trying the setup account page:
Screenshot 2023-10-20 at 10 22 12
Also @madewulf i'm not getting the point of displaying modules list without any action on this page:

Screenshot 2023-10-20 at 10 20 29

**Comment by hakifran** [comment]
> I have a permission error while trying the setup account page: Screenshot 2023-10-20 at 10 22 12 Also @madewulf i'm not getting the point of displaying modules list without any action on this page:
> Screenshot 2023-10-20 at 10 20 29

@beygorghor for the question about the displaying modules list, I added it regarding this mockup https://whimsical.com/iaso-admin-screen-6hGzbQA4F2AKAEu8wFGL8y

**Comment by hakifran** [comment]
> I have a permission error while trying the setup account page: Screenshot 2023-10-20 at 10 22 12 Also @madewulf i'm not getting the point of displaying modules list without any action on this page:
> Screenshot 2023-10-20 at 10 20 29

@beygorghor for the error that you don't have the permission, are you using a superUser? if not the user you are using should have iaso_modules permission

File: hat/assets/js/apps/Iaso/domains/app/translations/fr.json
Comment (hakifran) [reply hat/assets/js/apps/Iaso/domains/app/translations/fr.json:1]: @beygorghor pour les codename j'ai gardé ainsi parce que dans le code on les utilise tel qu'ils sont. Mais je vais les traduire en fr ou bien on les enleve simplement de l'affichage?

**Comment by beygorghor** [comment]
> > I have a permission error while trying the setup account page: Screenshot 2023-10-20 at 10 22 12 Also @madewulf i'm not getting the point of displaying modules list without any action on this page:
> > Screenshot 2023-10-20 at 10 20 29
> 
> @beygorghor for the error that you don't have the permission, are you using a superUser? if not the user you are using should have iaso_modules permission

@hakifran  The user is a user without profile then without permission. It's a staff user status, not super user, but staff can access this page too.
Screenshot 2023-10-24 at 11 40 45

**Review by beygorghor — APPROVED** [review]
Except the staff user error, it's ok for me

File: iaso/api/permissions.py [hunk 7]
```diff
@@ -28,7 +29,18 @@ def list(request):
...
+        account = request.user.iaso_profile.account
+        account_modules = account.modules if account.modules else []
+
+        # Get all permissions linked to the modules
+        modules_permissions = []
+        modules = MODULE_PERMISSIONS.keys()
+
+        for module in account_modules:
+            if module in modules:
+                modules_permissions = modules_permissions + MODULE_PERMISSIONS[module]
```
Comment (madewulf) [inline iaso/api/permissions.py:41]: Toute petite remarque, pas bloquante: ici, tu pourrais un jour avoir deux modules qui incluent la même permission, et dans ce cas, tu pourrais finir avec la même permission qui appraîtrait 2 fois dans modules_permissions. Je ne pense pas que ça pose problème plus loin (tu fais _in=modules_permission) mais j'aurais eu tendance à utiliser un set plutôt qu'une liste pour éviter les doublons.

File: iaso/api/setup_account.py [hunk 8]
```diff
@@ -70,8 +91,18 @@ def create(self, validated_data):
         data_source.save()
 
         Profile.objects.create(account=account, user=user)
+        # Get all permissions linked to the modules
+        modules_permissions = []
+        modules = MODULE_PERMISSIONS.keys()
+
+        for module in account_modules:
```
Comment (madewulf) [inline iaso/api/setup_account.py:98]: Ce code est beaucoup répété, ça aurait sans doute mérité une petite fonction utilitaire

**Comment by madewulf** [comment]
> > > I have a permission error while trying the setup account page: Screenshot 2023-10-20 at 10 22 12 Also @madewulf i'm not getting the point of displaying modules list without any action on this page:
> > > Screenshot 2023-10-20 at 10 20 29
> > 
> > 
> > @beygorghor for the error that you don't have the permission, are you using a superUser? if not the user you are using should have iaso_modules permission
> 
> @hakifran The user is a user without profile then without permission. It's a staff user status, not super user, but staff can access this page too. Screenshot 2023-10-24 at 11 40 45

For that point, I tend to agree that it's not super useful, but I think we can keep it and just add a small explanation. 

Something like: "You can see here the feature modules that are activated for your account. Contact an admin of your instance if you wish to activate additional features" at the top.

**Pull request merged**